cpp/src/arrow/util/time.h

```diff
@@ -0,0 +1,57 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <utility>
+
+#include "arrow/result.h"
+#include "arrow/type.h"
+
+namespace arrow {
+namespace util {
+
+enum DivideOrMultiply {
+  MULTIPLY,
+  DIVIDE,
+};
+
+// TimestampType -> TimestampType
+static std::pair<DivideOrMultiply, int64_t> kTimestampConversionTable[4][4] = {
```

Comment:
What do you expose those values publicly and make the variable static, rather than extern? Is it out of hope that the compiler may conduct optimizations? (if so, you probably want to make it `const` or `constexpr` too)

Reply:
I originally did it with constexpr, but constexpr pair constructor is only supported in C++14, some compiler complained.

Reply:
We can recover constexpr with a custom struct